tensorflow/compiler/tf2tensorrt/segment/segment.cc

```diff
@@ -1141,6 +1143,76 @@ Status SegmentGraph(const Graph* tf_graph,
       continue;
     }
     segments->emplace_back(itr.second.property, segment_nodes);
+    effective_nodes_counts.push_back(num_effective_nodes);
+  }
+
+  // --------------------------------- Step 4 ---------------------------------
+  // If the number of segments exceeds max_engines, prune the smallest ones.
+
+  int64_t max_trt_engine_ops;
+  TF_CHECK_OK(ReadInt64FromEnvVar("TF_TRT_MAX_ALLOWED_ENGINES",
+                                  /*default_value=*/20, &max_trt_engine_ops));
+
+  if (max_trt_engine_ops <= 0) {
+    LOG(WARNING) << "The environment variable TF_TRT_MAX_ALLOWED_ENGINES is "
+                 << "<= 0. TF-TRT did not limit the number of TensorRT engines "
+                 << "created.";
+
+  } else {
+    if (segments->size() > max_trt_engine_ops) {
+      LOG(WARNING) << "A total of " << segments->size() << " segments with at "
+                   << "least minimum_segment_size="
+                   << options.minimum_segment_size << " nodes have been found. "
+                   << "TF-TRT will only convert the " << max_trt_engine_ops
+                   << " largest segments. You can change this behavior by "
+                   << "modifying the environment variable "
+                   << "TF_TRT_MAX_ALLOWED_ENGINES=" << max_trt_engine_ops;
+
+      // Stable sort of the segment indices according to their effective sizes.
+      std::vector<int> indices(segments->size());
+      std::iota(indices.begin(), indices.end(), 0);
+
+      std::stable_sort(indices.begin(), indices.end(),
+                      [&effective_nodes_counts](int i1, int i2) {
+                        return effective_nodes_counts[i1] >
+                                effective_nodes_counts[i2];
+                      });
+
+      // Log which segments have been accepted or rejected
+      VLOG(1) << "The following segments have been accepted by TF-TRT:";
+      for (int i = 0; i < indices.size() && i < max_trt_engine_ops; i++) {
+        auto segment_id = indices[i];
+        VLOG(1) << "- Segment " << segment_id << " [node count: "
+                << effective_nodes_counts[segment_id] << "]"
+                << " accepted. Re-assigned Segment ID=" << i;
```

Comment:
I don't think that these logs are correct. The array `indices` is sorted by effective node counts, but the code that I wrote with a mask keeps the original order of the segments.

Let's take a simple example:
```
segment 0: 50 effective nodes
segment 1: 30
segment 2: 100
max_trt_engine_ops=2
```
What actually happens is:
```
segment 0 -> segment 0
segment 1 discarded
segment 2 -> segment 1
```
What your logs will report is:
```
Segment 2 [node count: 100] accepted. Re-assigned Segment ID=0
Segment 0 [node count: 50] accepted. Re-assigned Segment ID=1
Segment 1 [node count: 30] rejected.
```

There are two ways to fix this:

A. Instead of using a mask like in my approach, use the indices directly to construct the new array. You lose the original ordering and you also need a temporary array. Something like (disclaimer: untested):
```c++
std::vector<Segment> temp;
for (int i = 0; i < max_trt_engine_ops; i++) {
    temp.push_back(segments->at(indices[i]));
}
segments->resize(max_trt_engine_ops);
std::copy(temp.begin(), temp.end(), segments->begin());
```

B. Keep the current behavior and fix the logs. E.g:
```c++
// Gather the masked elements at the start of the array, in place.
int j = 0;
for (int i = 0; i < segments->size(); i++) {
    if(mask[i]) {
        VLOG(1) << "Segment " << i << " [node count: " 
            << effective_nodes_counts[i] 
            << "] accepted. Re-assigned segment ID=" << j;
        segments->at(j) = segments->at(i);
        j++;
    }
}
for (int i = 0; i < segments->size(); i++) {
    if(!mask[i]) {
        VLOG(1) << "Segment " << i << " [node count: " 
            << effective_nodes_counts[i] << "] rejected.";
    }
}
```

Reply:
Good catch, Louis!
I support approach B. which keep the current behavior, so that this PR won't change the resulting engines if the model has less than 20 engines.

Reply:
Fixed